File: pulse/android/app/src/main/kotlin/com/example/pulse/BleAdvertiserPlugin.kt
package com.example.pulse

import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.util.Log
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.UUID

/**
 * Complete BLE Peripheral Plugin for Pulse Mesh Networking.
 * 
 * This plugin provides:
 * - BLE Advertising (to be discoverable)
 * - GATT Server (to receive incoming data from Central devices)
 * - EventChannel (to stream incoming data to Dart)
 */
class BleAdvertiserPlugin(private val context: Context) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    companion object {
        private const val TAG = "BleAdvertiser"
        
        // Service and Characteristic UUIDs - must match iOS and Dart code
        val SERVICE_UUID: UUID = UUID.fromString("0000FADE-0000-1000-8000-00805F9B34FB")
        val MESSAGE_CHAR_UUID: UUID = UUID.fromString("0000FAB0-0000-1000-8000-00805F9B34FB")
        val TYPING_CHAR_UUID: UUID = UUID.fromString("0000FAB1-0000-1000-8000-00805F9B34FB")
    }

    private var bluetoothManager: BluetoothManager? = null
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothLeAdvertiser: BluetoothLeAdvertiser? = null
    private var bluetoothGattServer: BluetoothGattServer? = null
    private var advertiseCallback: AdvertiseCallback? = null
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val connectedDevices = mutableSetOf<BluetoothDevice>()
    
    // Characteristics references for notifications
    private var messageCharacteristic: BluetoothGattCharacteristic? = null
    private var typingCharacteristic: BluetoothGattCharacteristic? = null

    // ==================== MethodChannel Handler ====================

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAdvertisingSupported" -> {
                result.success(isAdvertisingSupported())
            }
            "startAdvertising" -> {
                val serviceUuid = call.argument<String>("serviceUuid")
                val userId = call.argument<String>("userId")
                
                if (serviceUuid == null || userId == null) {
                    result.error("INVALID_ARGS", "serviceUuid and userId are required", null)
                    return
                }
                
                startAdvertising(serviceUuid, userId, result)
            }
            "stopAdvertising" -> {
                stopAdvertising()
                result.success(null)
            }
            "sendToConnectedDevices" -> {
                val data = call.argument<ByteArray>("data")
                val charType = call.argument<String>("characteristicType") ?: "message"
                if (data != null) {
                    sendToConnectedDevices(data, charType)
                    result.success(true)
                } else {
                    result.error("INVALID_ARGS", "data is required", null)
                }
            }
            "getConnectedDeviceCount" -> {
                result.success(connectedDevices.size)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // ==================== EventChannel StreamHandler ====================

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        Log.d(TAG, "EventChannel: Dart listener attached")
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        Log.d(TAG, "EventChannel: Dart listener detached")
        eventSink = null
    }

    // ==================== BLE Advertising ====================

    private fun isAdvertisingSupported(): Boolean {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        val bluetoothAdapter = bluetoothManager?.adapter
        return bluetoothAdapter?.bluetoothLeAdvertiser != null &&
               bluetoothAdapter.isMultipleAdvertisementSupported
    }

    private fun startAdvertising(serviceUuid: String, userId: String, result: MethodChannel.Result) {
        try {
            bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
            bluetoothAdapter = bluetoothManager?.adapter

            if (bluetoothAdapter == null || !bluetoothAdapter!!.isEnabled) {
                result.error("BT_DISABLED", "Bluetooth is not enabled", null)
                return
            }

            bluetoothLeAdvertiser = bluetoothAdapter!!.bluetoothLeAdvertiser
            if (bluetoothLeAdvertiser == null) {
                result.error("NOT_SUPPORTED", "BLE advertising not supported on this device", null)
                return
            }

            // Parse UUID (use our standard service UUID)
            val uuid = try {
                UUID.fromString(serviceUuid)
            } catch (e: IllegalArgumentException) {
                result.error("INVALID_UUID", "Invalid service UUID format", null)
                return
            }

            // First, setup the GATT server
            setupGattServer()

            // Configure advertising settings
            val settings = AdvertiseSettings.Builder()
                .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
                .setConnectable(true)
                .setTimeout(0) // Advertise indefinitely
                .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_HIGH)
                .build()

            // Configure advertising data
            val data = AdvertiseData.Builder()
                .setIncludeDeviceName(false)
                .addServiceUuid(ParcelUuid(uuid))
                .build()

            // Configure scan response
            val scanResponse = AdvertiseData.Builder()
                .setIncludeDeviceName(true)
                .build()

            // Create callback
            advertiseCallback = object : AdvertiseCallback() {
                override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
                    Log.d(TAG, "✅ BLE advertising started successfully")
                    mainHandler.post { result.success(true) }
                }

                override fun onStartFailure(errorCode: Int) {
                    val errorMsg = when (errorCode) {
                        ADVERTISE_FAILED_DATA_TOO_LARGE -> "Data too large"
                        ADVERTISE_FAILED_TOO_MANY_ADVERTISERS -> "Too many advertisers"
                        ADVERTISE_FAILED_ALREADY_STARTED -> "Already started"
                        ADVERTISE_FAILED_INTERNAL_ERROR -> "Internal error"
                        ADVERTISE_FAILED_FEATURE_UNSUPPORTED -> "Feature unsupported"
                        else -> "Unknown error: $errorCode"
                    }
                    Log.e(TAG, "❌ BLE advertising failed: $errorMsg")
                    mainHandler.post { result.error("ADVERTISE_FAILED", errorMsg, errorCode) }
                }
            }

            // Start advertising
            bluetoothLeAdvertiser?.startAdvertising(settings, data, scanResponse, advertiseCallback!!)

        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception - missing Bluetooth permissions", e)
            result.error("PERMISSION_DENIED", "Bluetooth permissions not granted", null)
        } catch (e: Exception) {
            Log.e(TAG, "Error starting BLE advertising", e)
            result.error("EXCEPTION", e.message, null)
        }
    }

    private fun stopAdvertising() {
        try {
            advertiseCallback?.let { callback ->
                bluetoothLeAdvertiser?.stopAdvertising(callback)
                Log.d(TAG, "🛑 BLE advertising stopped")
            }
            advertiseCallback = null
            bluetoothLeAdvertiser = null
            
            // Close GATT server
            bluetoothGattServer?.close()
            bluetoothGattServer = null
            connectedDevices.clear()
            
        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception stopping advertising", e)
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping BLE advertising", e)
        }
    }

    // ==================== GATT Server ====================

    private fun setupGattServer() {
        try {
            bluetoothGattServer = bluetoothManager?.openGattServer(context, gattServerCallback)
            
            if (bluetoothGattServer == null) {
                Log.e(TAG, "❌ Failed to open GATT server")
                return
            }

            // Create the service
            val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)

            // Create message characteristic (read, write, notify)
            messageCharacteristic = BluetoothGattCharacteristic(
                MESSAGE_CHAR_UUID,
                BluetoothGattCharacteristic.PROPERTY_READ or
                BluetoothGattCharacteristic.PROPERTY_WRITE or
                BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE or
                BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_READ or
                BluetoothGattCharacteristic.PERMISSION_WRITE
            )
            
            // Add Client Characteristic Configuration Descriptor for notifications
            val messageDescriptor = BluetoothGattDescriptor(
                UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"), // CCCD UUID
                BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE
            )
            messageCharacteristic!!.addDescriptor(messageDescriptor)
            service.addCharacteristic(messageCharacteristic!!)

            // Create typing characteristic (read, write, notify)
            typingCharacteristic = BluetoothGattCharacteristic(
                TYPING_CHAR_UUID,
                BluetoothGattCharacteristic.PROPERTY_READ or
                BluetoothGattCharacteristic.PROPERTY_WRITE or
                BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE or
                BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_READ or
                BluetoothGattCharacteristic.PERMISSION_WRITE
            )
            
            val typingDescriptor = BluetoothGattDescriptor(
                UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"),
                BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE
            )
            typingCharacteristic!!.addDescriptor(typingDescriptor)
            service.addCharacteristic(typingCharacteristic!!)

            // Add service to GATT server
            val added = bluetoothGattServer?.addService(service) ?: false
            if (added) {
                Log.d(TAG, "✅ GATT service added successfully")
            } else {
                Log.e(TAG, "❌ Failed to add GATT service")
            }

        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception setting up GATT server", e)
        } catch (e: Exception) {
            Log.e(TAG, "Error setting up GATT server", e)
        }
    }

    private val gattServerCallback = object : BluetoothGattServerCallback() {
        
        override fun onConnectionStateChange(device: BluetoothDevice?, status: Int, newState: Int) {
            super.onConnectionStateChange(device, status, newState)
            try {
                if (device == null) return
                
                when (newState) {
                    BluetoothProfile.STATE_CONNECTED -> {
                        connectedDevices.add(device)
                        Log.d(TAG, "🔵 Device connected: ${device.address} (${connectedDevices.size} total)")
                        sendEventToDart(mapOf(
                            "type" to "connection",
                            "event" to "connected",
                            "deviceAddress" to device.address,
                            "connectedCount" to connectedDevices.size
                        ))
                    }
                    BluetoothProfile.STATE_DISCONNECTED -> {
                        connectedDevices.remove(device)
                        Log.d(TAG, "🔴 Device disconnected: ${device.address} (${connectedDevices.size} remaining)")
                        sendEventToDart(mapOf(
                            "type" to "connection",
                            "event" to "disconnected",
                            "deviceAddress" to device.address,
                            "connectedCount" to connectedDevices.size
                        ))
                    }
                }
            } catch (e: SecurityException) {
                Log.e(TAG, "Security exception in connection state change", e)
            }
        }

        override fun onCharacteristicReadRequest(
            device: BluetoothDevice?,
            requestId: Int,
            offset: Int,
            characteristic: BluetoothGattCharacteristic?
        ) {
            super.onCharacteristicReadRequest(device, requestId, offset, characteristic)
            try {
                Log.d(TAG, "📖 Read request from ${device?.address} for ${characteristic?.uuid}")
                bluetoothGattServer?.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    offset,
                    characteristic?.value ?: ByteArray(0)
                )
            } catch (e: SecurityException) {
                Log.e(TAG, "Security exception handling read request", e)
            }
        }

        override fun onCharacteristicWriteRequest(
            device: BluetoothDevice?,
            requestId: Int,
            characteristic: BluetoothGattCharacteristic?,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray?
        ) {
            super.onCharacteristicWriteRequest(device, requestId, characteristic, preparedWrite, responseNeeded, offset, value)
            try {
                val charUuid = characteristic?.uuid
                val dataSize = value?.size ?: 0
                Log.d(TAG, "📥 Write request: $dataSize bytes from ${device?.address} to $charUuid")

                // Send response if needed
                if (responseNeeded) {
                    bluetoothGattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        offset,
                        value
                    )
                }

                // Forward data to Dart via EventChannel
                if (value != null && value.isNotEmpty()) {
                    val charType = when (charUuid) {
                        MESSAGE_CHAR_UUID -> "message"
                        TYPING_CHAR_UUID -> "typing"
                        else -> "unknown"
                    }
                    
                    sendEventToDart(mapOf(
                        "type" to "data",
                        "characteristicType" to charType,
                        "data" to value.toList(), // Convert to List<Int> for Dart
                        "deviceAddress" to (device?.address ?: "unknown")
                    ))
                }

            } catch (e: SecurityException) {
                Log.e(TAG, "Security exception handling write request", e)
            } catch (e: Exception) {
                Log.e(TAG, "Error handling write request", e)
            }
        }

        override fun onDescriptorWriteRequest(
            device: BluetoothDevice?,
            requestId: Int,
            descriptor: BluetoothGattDescriptor?,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray?
        ) {
            super.onDescriptorWriteRequest(device, requestId, descriptor, preparedWrite, responseNeeded, offset, value)
            try {
                Log.d(TAG, "📝 Descriptor write from ${device?.address}")
                
                // Handle CCCD writes for notifications
                if (responseNeeded) {
                    bluetoothGattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        offset,
                        value
                    )
                }
            } catch (e: SecurityException) {
                Log.e(TAG, "Security exception handling descriptor write", e)
            }
        }

        override fun onServiceAdded(status: Int, service: BluetoothGattService?) {
            super.onServiceAdded(status, service)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(TAG, "✅ Service ${service?.uuid} added successfully")
            } else {
                Log.e(TAG, "❌ Failed to add service: status $status")
            }
        }

        override fun onNotificationSent(device: BluetoothDevice?, status: Int) {
            super.onNotificationSent(device, status)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(TAG, "✅ Notification sent to ${device?.address}")
            } else {
                Log.e(TAG, "❌ Notification failed to ${device?.address}: status $status")
            }
        }
    }

    // ==================== Send Data to Connected Devices ====================

    private fun sendToConnectedDevices(data: ByteArray, charType: String) {
        try {
            val characteristic = when (charType) {
                "typing" -> typingCharacteristic
                else -> messageCharacteristic
            }

            if (characteristic == null) {
                Log.e(TAG, "Characteristic not available for type: $charType")
                return
            }

            characteristic.value = data

            for (device in connectedDevices.toList()) {
                try {
                    val sent = bluetoothGattServer?.notifyCharacteristicChanged(
                        device,
                        characteristic,
                        false // indicate = false means notification
                    )
                    Log.d(TAG, "📤 Notified ${device.address}: $sent (${data.size} bytes)")
                } catch (e: SecurityException) {
                    Log.e(TAG, "Security exception notifying device", e)
                } catch (e: Exception) {
                    Log.e(TAG, "Error notifying device ${device.address}", e)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sending to connected devices", e)
        }
    }

    // ==================== Event Sending ====================

    private fun sendEventToDart(data: Map<String, Any>) {
        mainHandler.post {
            eventSink?.success(data)
        }
    }

    // ==================== Cleanup ====================

    fun cleanup() {
        stopAdvertising()
        eventSink = null
    }
}

File: pulse/ios/Runner/BleAdvertiserPlugin.swift
import Flutter
import CoreBluetooth
import UIKit

/**
 * Complete BLE Peripheral Plugin for Pulse Mesh Networking (iOS).
 *
 * This plugin provides:
 * - BLE Advertising (to be discoverable)
 * - GATT Server via CBPeripheralManager (to receive incoming data)
 * - EventChannel (to stream incoming data to Dart)
 */
class BleAdvertiserPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, CBPeripheralManagerDelegate {
    
    // MARK: - Properties
    
    private var peripheralManager: CBPeripheralManager?
    private var isAdvertising = false
    private var serviceUUID: CBUUID?
    private var eventSink: FlutterEventSink?
    
    // Characteristic UUIDs - must match Android and Dart code
    private let messageCharUUID = CBUUID(string: "0000FAB0-0000-1000-8000-00805F9B34FB")
    private let typingCharUUID = CBUUID(string: "0000FAB1-0000-1000-8000-00805F9B34FB")
    
    // Mutable characteristics for sending notifications
    private var messageCharacteristic: CBMutableCharacteristic?
    private var typingCharacteristic: CBMutableCharacteristic?
    
    // Track subscribed centrals
    private var subscribedCentrals: [CBCentral] = []
    
    // Pending result for async operations
    private var pendingStartResult: FlutterResult?
    
    // MARK: - Plugin Registration
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = BleAdvertiserPlugin()
        
        // Method Channel for commands
        let methodChannel = FlutterMethodChannel(
            name: "com.pulse.ble/advertiser",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        
        // Event Channel for incoming BLE data
        let eventChannel = FlutterEventChannel(
            name: "com.pulse.ble/advertiser_events",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }
    
    // MARK: - FlutterStreamHandler
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("📡 iOS: EventChannel listener attached")
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("📡 iOS: EventChannel listener detached")
        self.eventSink = nil
        return nil
    }
    
    // MARK: - Method Channel Handler
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAdvertisingSupported":
            // iOS always supports BLE peripheral mode (if BLE is available)
            result(true)
            
        case "startAdvertising":
            guard let args = call.arguments as? [String: Any],
                  let serviceUuidString = args["serviceUuid"] as? String,
                  let userId = args["userId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "serviceUuid and userId are required", details: nil))
                return
            }
            startAdvertising(serviceUuid: serviceUuidString, userId: userId, result: result)
            
        case "stopAdvertising":
            stopAdvertising()
            result(nil)
            
        case "sendToConnectedDevices":
            guard let args = call.arguments as? [String: Any],
                  let dataList = args["data"] as? [Int] else {
                result(FlutterError(code: "INVALID_ARGS", message: "data is required", details: nil))
                return
            }
            let charType = args["characteristicType"] as? String ?? "message"
            let data = Data(dataList.map { UInt8($0) })
            sendToSubscribedCentrals(data: data, charType: charType)
            result(true)
            
        case "getConnectedDeviceCount":
            result(subscribedCentrals.count)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - BLE Advertising
    
    private func startAdvertising(serviceUuid: String, userId: String, result: @escaping FlutterResult) {
        guard let uuid = UUID(uuidString: serviceUuid) else {
            result(FlutterError(code: "INVALID_UUID", message: "Invalid service UUID format", details: nil))
            return
        }
        
        serviceUUID = CBUUID(nsuuid: uuid)
        pendingStartResult = result
        
        // Initialize peripheral manager if needed
        if peripheralManager == nil {
            peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
        } else if peripheralManager?.state == .poweredOn {
            actuallyStartAdvertising()
        }
        // If not powered on, will be called from peripheralManagerDidUpdateState
    }
    
    private func actuallyStartAdvertising() {
        guard let peripheralManager = peripheralManager,
              let serviceUUID = serviceUUID else {
            pendingStartResult?(FlutterError(code: "NOT_INITIALIZED", message: "Peripheral manager not initialized", details: nil))
            pendingStartResult = nil
            return
        }
        
        // Create service
        let service = CBMutableService(type: serviceUUID, primary: true)
        
        // Create message characteristic (read, write, notify)
        messageCharacteristic = CBMutableCharacteristic(
            type: messageCharUUID,
            properties: [.read, .write, .writeWithoutResponse, .notify],
            value: nil,
            permissions: [.readable, .writeable]
        )
        
        // Create typing characteristic (read, write, notify)
        typingCharacteristic = CBMutableCharacteristic(
            type: typingCharUUID,
            properties: [.read, .write, .writeWithoutResponse, .notify],
            value: nil,
            permissions: [.readable, .writeable]
        )
        
        service.characteristics = [messageCharacteristic!, typingCharacteristic!]
        
        // Add service
        peripheralManager.add(service)
        
        // Start advertising
        let advertisementData: [String: Any] = [
            CBAdvertisementDataServiceUUIDsKey: [serviceUUID],
            CBAdvertisementDataLocalNameKey: "Pulse"
        ]
        
        peripheralManager.startAdvertising(advertisementData)
        isAdvertising = true
        
        print("✅ iOS: BLE advertising started")
        pendingStartResult?(true)
        pendingStartResult = nil
    }
    
    private func stopAdvertising() {
        peripheralManager?.stopAdvertising()
        peripheralManager?.removeAllServices()
        isAdvertising = false
        subscribedCentrals.removeAll()
        print("🛑 iOS: BLE advertising stopped")
    }
    
    // MARK: - Send Data to Connected Centrals
    
    private func sendToSubscribedCentrals(data: Data, charType: String) {
        guard let peripheralManager = peripheralManager else {
            print("⚠️ iOS: PeripheralManager not available")
            return
        }
        
        let characteristic: CBMutableCharacteristic? = charType == "typing" ? typingCharacteristic : messageCharacteristic
        
        guard let char = characteristic else {
            print("⚠️ iOS: Characteristic not available for type: \(charType)")
            return
        }
        
        char.value = data
        
        // Send notification to all subscribed centrals
        let success = peripheralManager.updateValue(data, for: char, onSubscribedCentrals: nil)
        
        if success {
            print("📤 iOS: Sent \(data.count) bytes to \(subscribedCentrals.count) subscribed central(s)")
        } else {
            print("⚠️ iOS: Failed to send data (queue full), will retry")
            // The peripheral manager will call peripheralManagerIsReady(toUpdateSubscribers:) when ready
        }
    }
    
    // MARK: - CBPeripheralManagerDelegate
    
    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        var stateString = ""
        switch peripheral.state {
        case .poweredOn:
            stateString = "poweredOn"
            print("📱 iOS: Bluetooth powered on")
            // If we have a pending start request, proceed
            if pendingStartResult != nil {
                actuallyStartAdvertising()
            }
        case .poweredOff:
            stateString = "poweredOff"
            print("📱 iOS: Bluetooth powered off")
            isAdvertising = false
            subscribedCentrals.removeAll()
        case .resetting:
            stateString = "resetting"
            print("📱 iOS: Bluetooth resetting")
        case .unauthorized:
            stateString = "unauthorized"
            print("📱 iOS: Bluetooth unauthorized")
            pendingStartResult?(FlutterError(code: "UNAUTHORIZED", message: "Bluetooth not authorized", details: nil))
            pendingStartResult = nil
        case .unsupported:
            stateString = "unsupported"
            print("📱 iOS: Bluetooth unsupported")
            pendingStartResult?(FlutterError(code: "UNSUPPORTED", message: "BLE not supported", details: nil))
            pendingStartResult = nil
        case .unknown:
            stateString = "unknown"
            print("📱 iOS: Bluetooth state unknown")
        @unknown default:
            stateString = "unknown"
            print("📱 iOS: Bluetooth state unknown")
        }
        
        // Send state change event to Dart
        sendEventToDart([
            "type": "bluetoothState",
            "state": stateString
        ])
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
        if let error = error {
            print("❌ iOS: Error adding service: \(error.localizedDescription)")
        } else {
            print("✅ iOS: Service added successfully: \(service.uuid)")
        }
    }
    
    func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
        if let error = error {
            print("❌ iOS: Error starting advertising: \(error.localizedDescription)")
        } else {
            print("✅ iOS: Started advertising successfully")
        }
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didSubscribeTo characteristic: CBCharacteristic) {
        print("🔵 iOS: Central \(central.identifier) subscribed to \(characteristic.uuid)")
        if !subscribedCentrals.contains(where: { $0.identifier == central.identifier }) {
            subscribedCentrals.append(central)
        }
        
        sendEventToDart([
            "type": "connection",
            "event": "subscribed",
            "centralId": central.identifier.uuidString,
            "characteristicUuid": characteristic.uuid.uuidString,
            "connectedCount": subscribedCentrals.count
        ])
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didUnsubscribeFrom characteristic: CBCharacteristic) {
        print("🔴 iOS: Central \(central.identifier) unsubscribed from \(characteristic.uuid)")
        subscribedCentrals.removeAll { $0.identifier == central.identifier }
        
        sendEventToDart([
            "type": "connection",
            "event": "unsubscribed",
            "centralId": central.identifier.uuidString,
            "characteristicUuid": characteristic.uuid.uuidString,
            "connectedCount": subscribedCentrals.count
        ])
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
        // Handle incoming write requests from central devices
        for request in requests {
            if let value = request.value {
                let charUuid = request.characteristic.uuid
                let charType: String
                
                if charUuid == messageCharUUID {
                    charType = "message"
                } else if charUuid == typingCharUUID {
                    charType = "typing"
                } else {
                    charType = "unknown"
                }
                
                print("📥 iOS: Received write: \(value.count) bytes from \(request.central.identifier) to \(charType)")
                
                // Forward data to Dart via EventChannel
                sendEventToDart([
                    "type": "data",
                    "characteristicType": charType,
                    "data": Array(value), // Convert Data to [UInt8] which becomes List<int> in Dart
                    "centralId": request.central.identifier.uuidString
                ])
            }
            
            // Send success response
            peripheral.respond(to: request, withResult: .success)
        }
    }
    
    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
        print("📖 iOS: Received read request from \(request.central.identifier)")
        
        // Return current value or empty data
        if let char = request.characteristic as? CBMutableCharacteristic {
            request.value = char.value ?? Data()
        }
        
        peripheral.respond(to: request, withResult: .success)
    }
    
    func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
        // Called when the peripheral manager is ready to send more updates
        // This is a good place to retry any pending notifications
        print("📡 iOS: Ready to update subscribers")
    }
    
    // MARK: - Event Sending
    
    private func sendEventToDart(_ data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(data)
        }
    }
}
